storage/gcsemu/raw_http_test.go

```diff
@@ -0,0 +1,227 @@
+				assert.Equal(t, http.StatusOK, rsp.StatusCode)
+
+				var attrs api.Object
+				err = json.NewDecoder(bytes.NewReader(body)).Decode(&attrs)
+				assert.NilError(t, err)
+				assert.Equal(t, name, attrs.Name)
+				assert.Equal(t, bh.Name, attrs.Bucket)
+				assert.Equal(t, uint64(len(v1)), attrs.Size)
+				assert.Equal(t, expectMetaGen, attrs.Metageneration)
+			},
+		},
+		{
+			name: "rawPatchMeta",
+			makeRequest: func(t *testing.T) *http.Request {
+				u := fmt.Sprintf("%s/storage/v1/b/%s/o/%s", url, bh.Name, name)
+				t.Logf(u)
+				req, err := http.NewRequest("PATCH", u, strings.NewReader(`{"metadata": {"type": "tabby"}}`))
+				assert.NilError(t, err)
+				req.Header.Set("Content-Type", "application/json")
+				return req
+			},
+			checkResponse: func(t *testing.T, rsp *http.Response) {
+				body, err := io.ReadAll(rsp.Body)
+				assert.NilError(t, err)
+				assert.Equal(t, http.StatusOK, rsp.StatusCode)
+
+				expectMetaGen++
+
+				var attrs api.Object
+				err = json.NewDecoder(bytes.NewReader(body)).Decode(&attrs)
+				assert.NilError(t, err)
+				assert.Equal(t, name, attrs.Name)
+				assert.Equal(t, bh.Name, attrs.Bucket)
+				assert.Equal(t, uint64(len(v1)), attrs.Size)
+				assert.Equal(t, expectMetaGen, attrs.Metageneration)
+				assert.Equal(t, "tabby", attrs.Metadata["type"])
+			},
+		},
+		{
+			name: "rawUpload",
+			makeRequest: func(t *testing.T) *http.Request {
+				u := fmt.Sprintf("%s/upload/storage/v1/b/%s/o?uploadType=media&name=%s", url, bh.Name, name2)
+				t.Logf(u)
+				req, err := http.NewRequest("POST", u, strings.NewReader(v2))
+				assert.NilError(t, err)
+				req.Header.Set("Content-Type", "text/plain")
+				return req
+			},
+			checkResponse: func(t *testing.T, rsp *http.Response) {
+				body, err := io.ReadAll(rsp.Body)
+				assert.NilError(t, err)
+				assert.Equal(t, http.StatusOK, rsp.StatusCode)
+
+				var attrs api.Object
+				err = json.NewDecoder(bytes.NewReader(body)).Decode(&attrs)
+				assert.NilError(t, err)
+				assert.Equal(t, name2, attrs.Name)
+				assert.Equal(t, bh.Name, attrs.Bucket)
+				assert.Equal(t, uint64(len(v2)), attrs.Size)
+				assert.Equal(t, int64(1), attrs.Metageneration)
+			},
+		},
+		{
+			name: "publicUrl",
+			makeRequest: func(t *testing.T) *http.Request {
+				u := fmt.Sprintf("%s/%s/%s?alt=media", url, bh.Name, name)
+				t.Logf(u)
+				req, err := http.NewRequest("GET", u, nil)
+				assert.NilError(t, err)
+				return req
+			},
+			checkResponse: func(t *testing.T, rsp *http.Response) {
+				body, err := io.ReadAll(rsp.Body)
+				assert.NilError(t, err)
+				assert.Equal(t, http.StatusOK, rsp.StatusCode)
+				assert.Equal(t, v1, string(body))
+			},
+		},
+	}
+
+	ctx := context.Background()
+	oh := bh.Object(name)
+
+	// Create the object 1.
+	w := oh.NewWriter(ctx)
+	assert.NilError(t, write(w, v1))
+
+	// Make sure object 2 is not there.
+	_ = bh.Object(name2).Delete(ctx)
+
+	// Run each test individually.
+	for _, tc := range tcs {
+		tc := tc
```

Comment:
Nit: Do we need to shadow this variable? I thought `t.Run` blocks while the test is running, unless we explicitly specify `t.Parallel()` inside it, so these tests will run sequentially, right?

Reply:
True, this was kind of defensive coding.